urlaeeaf948fe_smcpeak_ded_tgz-pJ8-ded_model_SerializationTestsJ8/src/util/awt/G.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    /** Draw 'image' in 'g' with upper-left at 'pt' and scaled to 'dim' (a Point). */
+    public static void drawImage(Graphics g, Image image, Point pt, Point dim)
+    {
+        drawImage(g, image, pt, toDimension(dim));
+    }
+
+    /** Draw 'image' in 'g' filling 'r'. */
+    public static void drawImage(Graphics g, Image image, Rectangle r)
+    {
+        drawImage(g, image, topLeft(r), r.getSize());
+    }
+
+    /** Convert a Point to a Dimension. */
+    public static Dimension toDimension(Point pt)
+    {
+        return new Dimension(pt.x, pt.y);
+    }
+
+    /** Convert a Dimension to a Point. */
+    public static Point toPoint(Dimension dim)
+    {
+        return new Point(dim.width, dim.height);
+    }
+
+    /** Subtract 'dim' from 'pt'. */
+    public static Point sub(Point pt, Dimension dim)
+    {
+        return new Point(pt.x - dim.width, pt.y - dim.height);
+    }
+
+    /** Return a new Point with x/y swapped. */
+    public static Point transpose(Point pt)
+    {
+        return new Point(pt.y, pt.x);
+    }
+
+    /** Return a new Point with x and y negated. */
+    public static Point negate(Point pt)
+    {
+        return new Point(-pt.x, -pt.y);
+    }
+
+    /** Return a new Rectangle based on 'r' where the top/left corner
+      * is moved by 'dist', leaving its bottom/left corner unchanged. */
+    public static Rectangle moveTopLeftBy(Rectangle r, Point dist)
+    {
+        return new Rectangle(
+            r.x + dist.x,
+            r.y + dist.y,
+            r.width - dist.x,
+            r.height - dist.y);
+    }
+
+    /** Return a new Rectangle based on 'r' where x or y, depending on 'hv',
+      * has been moved by 'delta', but the bottom/right corner is unchanged. */
+    public static Rectangle moveOriginBy(Rectangle r, HorizOrVert hv, int delta)
+    {
+        if (hv.isHoriz()) {
+            return new Rectangle(
+                r.x + delta,
+                r.y,
+                r.width - delta,
+                r.height);
+        }
+        else {
+            return new Rectangle(
+                r.x,
+                r.y + delta,
+                r.width,
+                r.height - delta);
+        }
+    }
+
+    /** Return a new Rectangle based on 'r' where the bottom/right corner
+      * is moved by 'dist', leaving its top/left corner unchanged. */
+    public static Rectangle moveBottomRightBy(Rectangle r, Point dist)
+    {
+        return new Rectangle(
+            r.x,
+            r.y,
+            r.width + dist.x,
+            r.height + dist.y);
+    }
+
+    /** Return a new Rectangle based on 'r' where its x or y, depending
+      * on 'hv', is set to 'nv', with size unchanged. */
+    public static Rectangle setOrigin(Rectangle r, HorizOrVert hv, int nv)
+    {
+        return new Rectangle(
+            hv.isHoriz()? nv : r.x,
+            hv.isVert()? nv : r.y,
+            r.width,
+            r.height);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 144:148, 162:166.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.